File: tools/record_turn_attachments.py
#!/usr/bin/env python3
"""Record the Python verdicts in suites/harness-turn-attachments/cases.json.

Runs prism-harness-py's real ``admit_attachments()`` on each case. The case file
is edited as TEXT, replacing each ``"py": ...`` verdict in place, so the recorder
cannot reformat or retype the rows it records.

    PRISM_HARNESS_PY=../prism-harness-py python tools/record_turn_attachments.py [--check]
"""

from __future__ import annotations

import base64
import json
import os
import sys
from pathlib import Path
from typing import Any

ROOT = Path(__file__).resolve().parent.parent
HARNESS = Path(os.environ.get("PRISM_HARNESS_PY", ROOT.parent / "prism-harness-py"))
sys.path.insert(0, str(HARNESS / "src"))

from prism_harness import HarnessError, admit_attachments  # noqa: E402


class _FromFile:
    """What prism-ai-core's from_local_path() gives a caller: media that knows it came
    from a file, and serializes as bytes with no path."""

    def __init__(self, payload: dict[str, Any]) -> None:
        self._payload = payload

    def to_dict(self) -> dict[str, Any]:
        return dict(self._payload)

    def is_url(self) -> bool:
        return False

    def is_file(self) -> bool:
        return True


def attachment_for(spec: dict[str, Any]) -> object:
    """The same spec, built the way a Python caller would hold it.

    Mirrored in prism-harness-py/tests/test_turn_attachments_corpus.py.
    """
    if spec["$"] == "Text":
        return {"text": spec["text"]}
    if spec["$"] == "String":
        return spec["value"]

    kind = "document" if spec["$"] == "Document" else "image"
    base: dict[str, Any] = {
        "kind": kind,
        "url": None,
        "base64": None,
        "mime_type": spec.get("mimeType"),
        "file_id": None,
        "filename": None,
    }

    def titled(payload: dict[str, Any]) -> dict[str, Any]:
        if kind != "document":
            return payload
        return {**payload, "document_title": spec.get("title"), "chunks": payload.get("chunks")}

    source = spec["from"]
    if source == "base64":
        return titled({**base, "base64": spec["base64"]})
    if source == "url":
        return titled({**base, "url": spec["url"]})
    if source == "urlWithBytes":
        return titled({**base, "url": spec["url"], "base64": spec["base64"]})
    if source == "localPath":
        encoded = base64.b64encode(spec["bytes"].encode("utf-8")).decode("ascii")
        return _FromFile(titled({**base, "base64": encoded}))
    if source == "fileId":
        return titled({**base, "file_id": spec["fileId"]})
    if source == "chunks":
        return titled({**base, "chunks": spec["chunks"]})
    if source == "text":
        encoded = base64.b64encode(spec["text"].encode("utf-8")).decode("ascii")
        return titled({**base, "base64": encoded, "mime_type": "text/plain"})
    if source == "nothing":
        return titled(base)
    raise ValueError(f"Unknown media source {source}")


def main() -> int:
    check = "--check" in sys.argv
    path = ROOT / "suites" / "harness-turn-attachments" / "cases.json"
    raw = path.read_text(encoding="utf-8")
    document = json.loads(raw)
    stale: list[str] = []

    for case in document["cases"]:
        try:
            admit_attachments(case["prompt"], [attachment_for(spec) for spec in case["attachments"]])
            verdict = "admitted"
        except HarnessError as refused:
            verdict = str(refused.code)

        if case["verdict"]["py"] != verdict:
            stale.append(case["id"])

        at = raw.index(f'"id": "{case["id"]}"')
        verdict_at = raw.index('"py": ', at)
        end = raw.index('"', verdict_at + len('"py": "'))
        raw = f'{raw[:verdict_at]}"py": "{verdict}{raw[end:]}'

    if check:
        if stale:
            print(f"Stale py verdicts: {', '.join(stale)}", file=sys.stderr)
            return 1
        print("PY verdicts current.", file=sys.stderr)
        return 0

    path.write_text(raw, encoding="utf-8", newline="\n")
    print(f"Wrote {len(document['cases'])} py verdict(s); {len(stale)} changed.", file=sys.stderr)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/record-turn-attachments.mjs
#!/usr/bin/env node
// Record the TypeScript verdicts in suites/harness-turn-attachments/cases.json,
// by running prism-harness-ts's real admitAttachments() on each case.
//
// The case file is edited as TEXT, replacing each `"ts": ...` verdict in place,
// so the recorder cannot reformat or retype the rows it records.
//
//   PRISM_HARNESS_TS=../prism-harness-ts node tools/record-turn-attachments.mjs [--check]

import { readFileSync, writeFileSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';

const root = dirname(dirname(fileURLToPath(import.meta.url)));
const harnessRoot = process.env.PRISM_HARNESS_TS ?? join(root, '..', 'prism-harness-ts');
const { admitAttachments, HarnessError } = await import(pathToFileURL(join(harnessRoot, 'dist', 'index.js')).href);

const check = process.argv.includes('--check');
const path = join(root, 'suites', 'harness-turn-attachments', 'cases.json');
let raw = readFileSync(path, 'utf8');
const document = JSON.parse(raw);

// The same spec, built the way a TypeScript caller would hold it. See the
// matching converter in prism-harness-ts/test/turn-attachments-corpus.test.ts.
export function attachmentFor(spec) {
  const kind = spec.$ === 'Document' ? 'document' : 'image';
  const base = { kind, url: null, base64: null, mime_type: spec.mimeType ?? null, file_id: null, filename: null };
  const titled = (object) => (kind === 'document' ? { ...object, document_title: spec.title ?? null, chunks: object.chunks ?? null } : object);

  switch (spec.$) {
    case 'Text':
      return { text: spec.text };
    case 'String':
      return spec.value;
  }

  switch (spec.from) {
    case 'base64':
      return titled({ ...base, base64: spec.base64 });
    case 'url':
      return titled({ ...base, url: spec.url });
    case 'urlWithBytes':
      return titled({ ...base, url: spec.url, base64: spec.base64 });
    case 'localPath':
      // What prism-ts's fromLocalPath() gives a caller: an object that knows it
      // came from a file, and serializes as bytes with no path.
      return {
        toObject: () => titled({ ...base, base64: Buffer.from(spec.bytes).toString('base64') }),
        isUrl: () => false,
        isFile: () => true,
      };
    case 'fileId':
      return titled({ ...base, file_id: spec.fileId });
    case 'chunks':
      return titled({ ...base, chunks: spec.chunks });
    case 'text':
      return titled({ ...base, base64: Buffer.from(spec.text).toString('base64'), mime_type: 'text/plain' });
    case 'nothing':
      return titled(base);
    default:
      throw new Error(`Unknown media source ${spec.from}`);
  }
}

const stale = [];

for (const testCase of document.cases) {
  let verdict;

  try {
    admitAttachments(testCase.prompt, testCase.attachments.map(attachmentFor));
    verdict = 'admitted';
  } catch (error) {
    if (!(error instanceof HarnessError)) throw error;
    verdict = error.code;
  }

  if (testCase.verdict.ts !== verdict) stale.push(testCase.id);

  const at = raw.indexOf(`"id": "${testCase.id}"`);
  const verdictAt = raw.indexOf('"ts": ', at);
  const end = raw.indexOf('"', verdictAt + '"ts": "'.length);
  raw = `${raw.slice(0, verdictAt)}"ts": "${verdict}${raw.slice(end)}`;
}

if (check) {
  if (stale.length > 0) {
    console.error(`Stale ts verdicts: ${stale.join(', ')}`);
    process.exit(1);
  }
  console.error('TS verdicts current.');
} else {
  writeFileSync(path, raw);
  console.error(`Wrote ${document.cases.length} ts verdict(s); ${stale.length} changed.`);
}
